lib/SILOptimizer/Differentiation/LinearMapInfo.cpp

```diff
@@ -331,10 +333,24 @@ void LinearMapInfo::generateDifferentiationDataStructures(
   }
 
   // Add linear map fields to the linear map tuples.
-  for (auto &origBB : *original) {
+  // Now we need to be very careful as we're having a very subtle
```

Comment:
Could you please explain why the reverse post-order traversal(RPOT) is needed and how it is helping? 

I understand that we need the branch trace enum for a BB to be fully constructed before using it to derive the type for the linear map tuple of the same BB. But I don't understand how going in RPOT order is going to ensure that we have the right types for the branch trace enums? 

I think so because while fully constructing a branch trace enum declaration in `populateBranchingTraceDecl` we look up the linear map tuple types of the predecessor BBs, but these types haven't been set yet.

Reply:
When traversing BBs in RPOT we ensure that each BB is processed *after* its predecessors. As a result, we know that all linear map tuples for predecessor BBs are already finalized and therefore the enum type that we created will also be "complete" – we will not need to add any entries later.

The problem was not linear map tuples, but that we passed "incomplete" `EnumDecl` to `getBranchingTraceEnumLoweredType`. As a result, several flags on the corresponding SIL type were set improperly, causing the branch trace enum type to be always trivial (despite non-trivial payloads added afterwards, the lowered type will be cached and not re-calculated).

Reply:
> When traversing BBs in RPOT we ensure that each BB is processed after its predecessors.

Ah gotcha. I was mistaking RPOT for regular post order traversal but (Right->Left->Node) instead of (Left->Right->Node). But it seems like it's more or less a "pre-order" traversal.

Reply:
No, it's neither a pre-order nor a post-order, it's not a DFS. We cannot use DFS because we're having a DAG, not a tree. Think about `A -> B -> C -> D; B->D` CFG. We need to visit `D` after *both* `B` and `C`. RPOT is quite a standard technique for various data-flow problems.

Reply:
That makes sense, thanks for the explanation!

Reply:
> The problem was not linear map tuples, but that we passed "incomplete" EnumDecl to getBranchingTraceEnumLoweredType. As a result, several flags on the corresponding SIL type were set improperly, causing the branch trace enum type to be always trivial (despite non-trivial payloads added afterwards, the lowered type will be cached and not re-calculated).

You're not supposed to mutate EnumDecls like that at all and changing the order of iteration is only papering over the issue. Type lowering caches the results because it assumes the inputs are immutable. You need to build your AST before you get to SIL.